### ISSUE TITLE
Define configuration format

### ISSUE THREAD

**Issue by lovesegfault** [Alexays/Waybar#434]
Right now the configuration file's format is a weird JSON variation with `//` comments, at least as I understand it.

It's be nice if we, instead, used a more config-friendly format (such as TOML or even YAML) that allowed for things like comments. Moreover, it's be useful if the file was called `config.json` or `config.toml` so that editors can pick-up the filetype automatically.

I'll leave a link to an article I enjoy on why JSON is a poor choice for a configuration language: https://www.lucidchart.com/techblog/2018/07/16/why-json-isnt-a-good-configuration-language/

**Comment by cafehaine** [comment]
That would mean breaking the compatibility with all installs of waybar. The user-base might be small, but this still would cause a lot of troubles.

**Comment by lovesegfault** [comment]
@cafehaine Not necessarily, the old config could remain, but be removed from wiki/readme and leave alongside the TOML one for some time. If you load the old config format it works with a deprecation warning.

A breaking change nonetheless, but with a nice adaption period.

**Comment by asmigala** [comment]
The great thing about yaml is that you can actually use a yaml parser to parse json. So that would kind of ensure backwards compatibility with older configs, the only problem being the `//` comments.

A dirty way to convert old config to yaml:
```
cat ~/.config/waybar/config | sed 's|//.*$||' | yq -y .
```

(this won't work if you have `//` in a string somewhere, but you get the idea)

**Comment by eoli3n** [comment]
+10 i'm really confused about that choice...

**Comment by eoli3n** [comment]
Toml is a bad choice too : [https://hitchdev.com/strictyaml/why-not/toml/](https://github.com/crdoconnor/strictyaml#why-strictyaml)
Yaml is still the easier and cleaner human readable configuration language and would be a great choice for that kind of project.

**Comment by lovesegfault** [comment]
@eoli3n Agree to disagree and I'll avoid bikeshedding this :laughing:

**Comment by WhyNotHugo** [comment]
FWIW, current format is just "a JSON variation", it's called `jsonc`, and actually used by plenty of tools. So I wouldn't discredit is as non-standard.

**Comment by eoli3n** [comment]
Any news ?

**Comment by billop** [comment]
> FWIW, current format is just "a JSON variation", it's called `jsonc`, and actually used by plenty of tools. So I wouldn't discredit is as non-standard.

I was only able to find that VS Code uses this format for configuration. What other tools were you talking about here?

**Comment by WhyNotHugo** [comment]
Off the top of my head, `nvim-coc` uses `jsonc`, as do flatpak manifests ([example](https://github.com/sonnyp/Tangram/blob/main/re.sonny.Tangram.json)). I guess `waybar` also counts on this list really. There's plenty of libraries for working with `jsonc` in various languages too.

I'm sure I've seen other applications, but my memory is bad.

Most of these tools use the `.json` extension, which is a bit annoying since editors are bad at figuring out it's actually `jsonc`.

I don't think popularity is that important though; it's very _easy_ to explain that `jsonc` is just "json with comments", and it's not like end users have to learn an entirely new format.

That said, I don't disagree with the initial criticism of JSON. But there's also a lot of valid criticism for YAML (e.g.: type handling). TOML is bad for highly nested structures, or when you have a large set of files that share a schema. For a configuration format like waybars, it's actually not a bad candidate.

Ultimately though, I think this falls into bikeshedding a bit. A PR showing pros and cons of a change is probably best at this point.

**Comment by liiil825** [comment]
I think it's not a matter of criticizing json or jsonc now, but adding support for yaml files is necessary, and the format of yaml is obviously more beautiful

**Comment by sascha-wi** [comment]
It boggles my mind that I can get syntax highlighting...

**Comment by xmalbertox** [comment]
> It boggles my mind that I can't get syntax highlighting...

May I ask which editor you use? For vim you can use the jsonc syntax highliting by installing [vim-polyglot](https://github.com/sheerun/vim-polyglot).

**Comment by wille** [comment]
+1 for yaml configuration

**Comment by sascha-wi** [comment]
> > It boggles my mind that I can't get syntax highlighting...
> 
> May I ask which editor you use? For vim you can use the jsonc syntax highliting by installing [vim-polyglot](https://github.com/sheerun/vim-polyglot).

Did you read this issue thread at all?

**Comment by yvvki** [comment]
I just wanted to ask if we can have a defined schema too, especially in the current JSON config. The schema can be included in with the `$schema` key (See <https://json-schema.org/>).

**Comment by Magniquick** [comment]
I made a schema: http://github.com/Magniquick/waybar-config-schema.
Warnings in the repo apply, tho it at least works for my config and the default config.

**Comment by muhammedaksam** [comment]
> I made a schema: [http://github.com/Magniquick/waybar-config-schema](https://github.com/Magniquick/waybar-config-schema). Warnings in the repo apply, tho it at least works for my config and the default config.

Thanks for the post! It’s actually a crazy coincidence—I was just in the middle of hacking together some workarounds for $schema when I found your reply from 12 hours ago.

**Comment by Magniquick** [comment]
> Thanks for the post! It’s actually a crazy coincidence—I was just in the middle of hacking together some workarounds for $schema when I found your reply from 12 hours ago.

:D thanks !
made some more changes and it's semi-stable atp (has automated strict pytests against mine and the default config). Do let me know if you have any ideas at all - The  current implementation scraps the waybar wiki and generates the schema from there (manually fixed some edge cases in code and some others by pushing fixes to the docs).

open issues if any module schemas dont work, and I'll take a look at them :)
